### PR TITLE
Added integer power to Numeric, along with property test

### PR DESCRIPTION
**Pull request by cjay**
Upstreamed from Liqwid.

### CODE REVIEW

**Review by kozross — CHANGES_REQUESTED** [review]
Cabal file needs a version bump.

**Comment by cjay** [comment]
Done, and added a newtype wrapper to prevent `PRational` reduction, since the original power function in Liqwid prevented reduction internally.

File: src/Plutarch/Extra/Numeric.hs [hunk 1]
```diff
@@ -10,3 +11,46 @@ peven ::
...
+ the 'Plutarch.Extra.PRationalNoReduce' wrapper to prevent reducing after each
+ internal multiplication.
+
+ @since 3.12.2
+-}
+(#^) ::
+  forall (s :: S) (n :: S -> Type) (e :: S -> Type).
+  (PNum n, PNum e, PEq e, PIntegral e) =>
+  Term s n ->
+  Term s e ->
```
Comment (kozross) [inline src/Plutarch/Extra/Numeric.hs:52]: Could we use `PPositive` for `e` here?

**Review by kozross — APPROVED** [review]

File: src/Plutarch/Extra/Numeric.hs (hunk 1 above)
Comment (cjay) [reply src/Plutarch/Extra/Numeric.hs:52]: Excluding 0 would restrict it too much, it would have to be `PNatural`. Should we make a `PNatural` in LPE?

Comment (kozross) [reply src/Plutarch/Extra/Numeric.hs:52]: Ah good point. I think for now we can leave it as `PIntegral`, as partiality in Plutarch isn't quite as bad as in Haskell.

Maybe `PNatural` would be a good thing for LPE in the future.

**Pull request merged**